### PR TITLE
Remove `PartialOrd`, `Ord` from `LocalDefId`

### PR DESCRIPTION
**Pull request by pierwill**
Part of work on https://github.com/rust-lang/rust/issues/90317.

### CODE REVIEW

**Review by cjgillot — COMMENTED** [review]
Thanks for the PR @pierwill! I left a few comments.
Those trait bounds on `LocalDefId` are used a lot, it may be easier to handle `Idx` and `PartialOrd/Ord` separately.
Changing data structures means we will have to be careful about perf implications.

**Comment by pierwill** [comment]
> Those trait bounds on LocalDefId are used a lot, it may be easier to handle Idx and PartialOrd/Ord separately.

Thanks for reviewing, @cjgillot! I'll work on just the ordering traits for now. I also might see if `DefId` is easier to start with (https://github.com/rust-lang/rust/issues/90317#issuecomment-955132184).

File: compiler/rustc_span/src/span_encoding.rs [hunk 1]
```diff
@@ -60,7 +60,7 @@ use rustc_data_structures::fx::FxIndexSet;
 /// the dependency to the parent definition's span. This is performed
 /// using the callback `SPAN_TRACK` to access the query engine.
 ///
-#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+#[derive(Clone, Copy, Eq, PartialEq, Hash, PartialOrd, Ord)]
```
[the remark this answers is not in the document]
Comment (pierwill) [reply compiler/rustc_span/src/span_encoding.rs:63]: I had a feeling this change was more than we wanted. 👍

[the remark this answers is not in the document]
Comment (pierwill) [reply compiler/rustc_span/src/span_encoding.rs:63]: > The former PartialOrd implementation using SpanData attempted to sort spans by position.

This is in historical code?

> Instead, we should keep an implementation in terms of SpanData, but stop comparing SpanData::parent

So will `Span` end up using manual implementations for `Ord` and `PartialOrd`?

**Comment by pierwill** [comment]
> Those trait bounds on LocalDefId are used a lot, it may be easier to handle Idx and PartialOrd/Ord separately.

Looks like `Idx` requires `Ord`...

File: compiler/rustc_span/src/lib.rs [hunk 2]
```diff
@@ -440,6 +440,18 @@ pub struct SpanData {
...
 
+impl Ord for SpanData {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.lo, self.hi, self.ctxt).cmp(&(other.lo, other.hi, other.ctxt))
+    }
+}
+
+impl PartialOrd for SpanData {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some((self.lo, self.hi, self.ctxt).cmp(&(other.lo, other.hi, other.ctxt)))
```
Comment (cjgillot) [inline compiler/rustc_span/src/lib.rs:451]:
```suggestion
        Some(self.cmp(other))
```

File: compiler/rustc_data_structures/src/graph/vec_graph/mod.rs [hunk 3]
```diff
@@ -20,7 +20,7 @@ pub struct VecGraph<N: Idx> {
 impl<N: Idx> VecGraph<N> {
     pub fn new(num_nodes: usize, mut edge_pairs: Vec<(N, N)>) -> Self {
         // Sort the edges by the source -- this is important.
-        edge_pairs.sort();
+        edge_pairs.sort_by_cached_key(|&edge_pairs| (edge_pairs.0.index(), edge_pairs.1.index()));
```
Comment (cjgillot) [inline compiler/rustc_data_structures/src/graph/vec_graph/mod.rs:23]: Why use the `cached_` version?

File: compiler/rustc_span/src/lib.rs [hunk 4]
```diff
@@ -440,6 +440,18 @@ pub struct SpanData {
     pub parent: Option<LocalDefId>,
 }
 
+impl Ord for SpanData {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.lo, self.hi, self.ctxt).cmp(&(other.lo, other.hi, other.ctxt))
```
Comment (cjgillot) [inline compiler/rustc_span/src/lib.rs:445]: Could you explicit destructure so we get a compile error when a field is added to `SpanData`? Can you add a comment on why we ignore `parent`?
`let SpanData { lo: s_lo, hi: s_hi, ctxt: s_ctxt, parent: _ } = self;` and likewise for `other`.

File: compiler/rustc_hir/src/hir_id.rs [hunk 5]
```diff
@@ -11,13 +12,29 @@ use std::fmt;
...
 /// the code base.
-#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug, PartialOrd, Ord)]
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
 #[derive(Encodable, Decodable)]
 pub struct HirId {
     pub owner: LocalDefId,
     pub local_id: ItemLocalId,
 }
 
+impl Ord for HirId {
```
Comment (cjgillot) [inline compiler/rustc_hir/src/hir_id.rs:22]: Can you remove the impls for PartialOrd/Ord for HirId too? It compares LocalDefIds, which we want to avoid.

**Review by cjgillot — COMMENTED** [review]
Thanks @pierwill. I added a few nits. The main remark is that we'd like to remove the PartialOrd/Ord impl for HirId, since it has exactly the same issue as the impl for LocalDefId.

File: compiler/rustc_data_structures/src/graph/vec_graph/mod.rs (hunk 3 above)
Comment (pierwill) [reply compiler/rustc_data_structures/src/graph/vec_graph/mod.rs:23]: No particular reason. I'd seen this method used over `sort_by_key` elsewhere in the compiler.

File: compiler/rustc_hir/src/hir_id.rs (hunk 5 above)
Comment (pierwill) [reply compiler/rustc_hir/src/hir_id.rs]: On it. 👍

File: compiler/rustc_index/src/vec.rs [hunk 6]
```diff
@@ -12,7 +12,7 @@ use std::vec;
 /// Represents some newtyped `usize` wrapper.
 ///
 /// Purpose: avoid mixing indexes for different bitvector domains.
-pub trait Idx: Copy + 'static + Ord + Debug + Hash {
+pub trait Idx: Copy + 'static + Eq + PartialEq + Debug + Hash {
```
Comment (pierwill) [inline compiler/rustc_index/src/vec.rs:15]: I don't remember why I had to do this... 🤔

Comment (pierwill) [reply compiler/rustc_index/src/vec.rs:15]: Ah, okay, it was needed in `rustc_data_structures::graph`.

**Comment by pierwill** [comment]
Going to need to locate the source of an infinite loop...

File: compiler/rustc_middle/src/mir/mono.rs [hunk 7]
```diff
@@ -366,7 +366,13 @@ impl<'tcx> CodegenUnit<'tcx> {
                             // instances into account. The others don't matter for
                             // the codegen tests and can even make item order
                             // unstable.
-                            InstanceDef::Item(def) => Some(def.did.index.as_usize()),
+                            InstanceDef::Item(def) => Some(
+                                def.did
+                                    .as_local()
+                                    .map(|def_id| tcx.hir().local_def_id_to_hir_id(def_id))
+                                    .unwrap()
```
Comment (pierwill) [inline compiler/rustc_middle/src/mir/mono.rs:373]: This will probably panic at some point...

**Comment by pierwill** [comment]
Huh. I did not expect this to affect any UI tests... https://github.com/rust-lang/rust/runs/4088478139?check_suite_focus=true#step:25:7148

File: compiler/rustc_middle/src/mir/query.rs [hunk 8]
```diff
@@ -358,13 +358,57 @@ pub enum ConstraintCategory {
...
+            _ => None,
+        };
+        let o = match other {
+            Self::ClosureUpvar(hir_id) => Some(hir_id.index()),
+            _ => None,
+        };
+
+        s.cmp(&o)
+    }
+}
```
Comment (pierwill) [inline compiler/rustc_middle/src/mir/query.rs:375]: This doesn't feel right to me...

File: compiler/rustc_middle/src/mir/query.rs [hunk 9]
```diff
@@ -358,13 +358,57 @@ pub enum ConstraintCategory {
...
+            &Self::ClosureUpvar(h) => Some(h.index()),
+        };
+        let o = match other {
+            &Self::Normal => None,
+            &Self::ClosureUpvar(h) => Some(h.index()),
+        };
+
+        s.cmp(&o)
+    }
+}
```
Comment (pierwill) [inline compiler/rustc_middle/src/mir/query.rs:404]: Same here.

Could these `Ord` implementations, if they're incorrect, be causing the changes in the UI tests?

File: compiler/rustc_data_structures/src/sorted_map/index_map.rs [hunk 10]
```diff
@@ -97,7 +97,7 @@ impl<I: Idx, K: Eq, V: Eq> Eq for SortedIndexMultiMap<I, K, V> {}
 impl<I: Idx, K: PartialEq, V: PartialEq> PartialEq for SortedIndexMultiMap<I, K, V> {
     fn eq(&self, other: &Self) -> bool {
         // No need to compare the sorted index. If the items are the same, the index will be too.
-        self.items == other.items
+        self == other
```
Comment (pierwill) [inline compiler/rustc_data_structures/src/sorted_map/index_map.rs:100]: Is this correct?

File: compiler/rustc_index/src/bit_set.rs [hunk 11]
```diff
@@ -590,13 +590,13 @@ impl<T: Idx> SparseBitSet<T> {
 
     fn contains(&self, elem: T) -> bool {
         assert!(elem.index() < self.domain_size);
-        self.elems.contains(&elem)
+        self.elems.iter().any(|e| elem.index() == e.index())
```
Comment (bjorn3) [inline compiler/rustc_index/src/bit_set.rs:593]: Are these changes necessary? `Eq` is still implemented, right?

Comment (pierwill) [reply compiler/rustc_index/src/bit_set.rs:593]: Ah! Let me check. I think there were some changes need as I was working to keep the build working, but not after completing certain changes. (This must be a common experience...)

File: compiler/rustc_data_structures/src/graph/vec_graph/mod.rs [hunk 12]
```diff
@@ -20,7 +20,7 @@ pub struct VecGraph<N: Idx> {
 impl<N: Idx> VecGraph<N> {
     pub fn new(num_nodes: usize, mut edge_pairs: Vec<(N, N)>) -> Self {
         // Sort the edges by the source -- this is important.
-        edge_pairs.sort();
+        edge_pairs.sort_by_key(|&edge_pairs| (edge_pairs.0.index(), edge_pairs.1.index()));
```
Comment (pierwill) [inline compiler/rustc_data_structures/src/graph/vec_graph/mod.rs:23]: Looks like this change was necessary.

File: compiler/rustc_span/src/lib.rs [hunk 13]
```diff
@@ -435,6 +435,33 @@ pub struct SpanData {
...
 }
 
+impl Ord for SpanData {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let SpanData {
+            lo: s_lo,
+            hi: s_hi,
+            ctxt: s_ctxt,
+            // `HirId` does not implement `Ord`.
+            parent: _,
```
[the remark this answers is not in the document]
Comment (pierwill) [reply compiler/rustc_span/src/lib.rs:445]: Now that I think of it, if we're [not removing](https://github.com/rust-lang/rust/pull/90408#issuecomment-963621886) `Ord` from `HirId` in this PR, this change isn't necessary.

File: compiler/rustc_span/src/lib.rs [hunk 14]
```diff
@@ -435,6 +435,33 @@ pub struct SpanData {
     pub parent: Option<LocalDefId>,
 }
 
+impl Ord for SpanData {
```
Comment (cjgillot) [inline compiler/rustc_span/src/lib.rs:438]: This should probably add a comment: "Order spans by position in the file."

File: compiler/rustc_span/src/lib.rs [hunk 15]
```diff
@@ -435,6 +435,33 @@ pub struct SpanData {
...
+    fn cmp(&self, other: &Self) -> Ordering {
+        let SpanData {
+            lo: s_lo,
+            hi: s_hi,
+            ctxt: s_ctxt,
+            // `HirId` does not implement `Ord`.
```
Comment (cjgillot) [inline compiler/rustc_span/src/lib.rs:444]:
```suggestion
            // `LocalDefId` does not implement `Ord`. Other fields are enough to determine in-file order.
```

File: compiler/rustc_span/src/lib.rs [hunk 16]
```diff
@@ -435,6 +435,33 @@ pub struct SpanData {
...
+            parent: _,
+        } = self;
+        let SpanData {
+            lo: o_lo,
+            hi: o_hi,
+            ctxt: o_ctxt,
+            // `HirId` does not implement `Ord`.
```
Comment (cjgillot) [inline compiler/rustc_span/src/lib.rs:451]:
```suggestion
            // `LocalDefId` does not implement `Ord`. Other fields are enough to determine in-file order.
```

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 17]
```diff
@@ -1300,7 +1300,7 @@ impl EncodeContext<'a, 'tcx> {
             })
             .collect::<Vec<_>>();
         // Sort everything to ensure a stable order for diagnotics.
-        keys_and_jobs.sort_by_key(|&(def_id, _, _)| def_id);
+        keys_and_jobs.sort_by_key(|&(def_id, _, _)| def_id.index());
```
Comment (cjgillot) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1303]: Should `mir_keys` eventually be an `IndexSet` to avoid the need for sorting?

**Review by cjgillot — APPROVED** [review]
Thanks @pierwill, LGTM. A few remaining nits and this will be ready for approval.

File: compiler/rustc_metadata/src/rmeta/encoder.rs (hunk 17 above)
Comment (pierwill) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1303]: Sounds good to me.

**Comment by pierwill** [comment]
Thanks, @cjgillot! Addressed your comments and rebased. 👍

**Comment by cjgillot** [comment]
Just in case, to avoid any surprise
@bors try @rust-timer queue

**Comment by pierwill** [comment]
> **Summary**: This change led to large relevant mixed results 🤷 in compiler performance.
> 
> * Large improvement in instruction counts (up to -2.9% on `full` builds of `deeply-nested-async`)
> * Large regression in instruction counts (up to 3.1% on `incr-unchanged` builds of `deeply-nested-async`)

So the effects are mixed, but limited to async heavy code. Nice that there's no regression in `incr-patched` builds, which seem like the most common cases for those developing async heavy code.

@rustbot label: +perf-regression-triaged

File: compiler/rustc_metadata/src/rmeta/encoder.rs (hunk 17 above)
Comment (pierwill) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1303]: This is a very simple. Maybe it's worth doing here. Could this have positive performance benefits?

Comment (pierwill) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1303]: @cjgillot For consideration: https://github.com/pierwill/rust/pull/1.

Comment (pierwill) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1303]: Opened https://github.com/rust-lang/rust/pull/90842.

**Comment by pierwill** [comment]
Looks like a rebase might fix this...

**Comment by cjgillot** [comment]
The failure seems legit. Probably just needs blessing the output.

**Comment by pierwill** [comment]
> 

Testing this now 👍

**Comment by pierwill** [comment]
> The failure seems legit. Probably just needs blessing the output.

Testing now 👍

**Comment by pierwill** [comment]
@cjgillot Maybe this will work? https://github.com/rust-lang/rust/pull/90408/commits/ab9dc25e22150a0a3a557a3f4da5d4356487b6fe

**Comment by pierwill** [comment]
New error has crept in. I'm having trouble making sense of this:

```
error[E0277]: can't compare `T` with `_`
   --> compiler/rustc_index/src/bit_set.rs:721:31
    |
721 |             if range.contains(e) {
    |                      -------- ^ no implementation for `T < _` and `T > _`
    |                      |
    |                      required by a bound introduced by this call
    |
note: required by a bound in `contains`
   --> /Users/wpierce/repos/rust/library/core/src/ops/range.rs:812:12
    |
812 |         T: PartialOrd<U>,
    |            ^^^^^^^^^^^^^ required by this bound in `contains`
```

**Comment by pierwill** [comment]
Having trouble understanding this error:

```
error[E0277]: can't compare `T` with `_`
   --> compiler/rustc_index/src/bit_set.rs:721:31
    |
721 |             if range.contains(e) {
    |                      -------- ^ no implementation for `T < _` and `T > _`
    |                      |
    |                      required by a bound introduced by this call
    |
note: required by a bound in `contains`
   --> /Users/u/repos/rust/library/core/src/ops/range.rs:812:12
    |
812 |         T: PartialOrd<U>,
    |            ^^^^^^^^^^^^^ required by this bound in `contains`
```

**Comment by pierwill** [comment]
This code was added in https://github.com/rust-lang/rust/commit/03afb61b53e27019166b8a2637a5eb7094017300. @Mark-Simulacrum: could offer some advice on troubleshooting this? 😇 

https://github.com/rust-lang/rust/runs/4433279740?check_suite_focus=true#step:25:509

**Comment by Mark-Simulacrum** [comment]
I'm not sure what aid I can offer, but it likely that the relevant method will want a where clause bound on having Ord, or something like that.

File: compiler/rustc_data_structures/src/graph/vec_graph/mod.rs (hunk 12 above)
Comment (Mark-Simulacrum) [inline compiler/rustc_data_structures/src/graph/vec_graph/mod.rs:23]: This seems a little concerning - I would rather expect that this data structure/method would require that the index type implements Ord. (Some types which are comparable, e.g. mir block indices, should continue to implement Ord, even as we remove it from DefId). Otherwise we're just hiding the actual Ord dependence which doesn't really help, imo.

**Comment by pierwill** [comment]
> Having trouble understanding this error:
> 
> ```
> error[E0277]: can't compare `T` with `_`
>    --> compiler/rustc_index/src/bit_set.rs:721:31
> [...]
> ```

What does it mean to compare with `_`? 😕

File: compiler/rustc_data_structures/src/graph/vec_graph/mod.rs (hunk 12 above)
Comment (pierwill) [reply compiler/rustc_data_structures/src/graph/vec_graph/mod.rs]: How does this look? https://github.com/rust-lang/rust/pull/90408/commits/44e023a64901a47348284b67b2c32c3817c1beb3

**Comment by pierwill** [comment]
Still working on some improvements in the spirit of this comment. @rustbot label -S-waiting-on-review

**Comment by pierwill** [comment]
@cjgillot I squashed one small commit just now, FYI. Not sure what caused that error...

**Comment by bjorn3** [comment]
It says that it was canceled. Probably spurious.

@bors retry

**Comment by pierwill** [comment]
@cjgillot Looks like bors isn't picking up the approval:

Screen Shot 2021-12-09 at 5 08 28 PM
 
My rebase probably invalidated it.

**Comment by pierwill** [comment]
This change has been failing the same Windows backtrace test as https://github.com/rust-lang/rust/pull/89836. We're [assessing](https://github.com/rust-lang/rust/pull/89836#issuecomment-992646395) whether to bless the test or continue digging.

**Comment by pierwill** [comment]
Once https://github.com/rust-lang/rust/pull/89836 is merged (🤞🤞), this should be ready to go. :)

**Comment by matthiaskrgr** [comment]
Just for the record, the failure here https://github.com/rust-lang/rust/pull/90408#issuecomment-991879747 is probably unrelated to this PR since it happened in other ci-runs as well :( :  https://github.com/rust-lang/rust/pull/91995#issuecomment-995710617

Made a tracking ticket: https://github.com/rust-lang/rust/issues/92000

**Comment by pierwill** [comment]
This is ready for another bors attempt! @cjgillot

**Comment by pierwill** [comment]
See previous comment here: https://github.com/rust-lang/rust/pull/90408#issuecomment-967448819.

@rustbot label: +perf-regression-triaged
